ccl/multiregionccl: TestMultiRegionDataDriven failed

ccl/multiregionccl.TestMultiRegionDataDriven [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4583447&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4583447&tab=artifacts#/) on release-22.1 @ [a1c1879e01ceee79a81693c67a1dba184b5fc1b1](https://github.com/cockroachdb/cockroach/commits/a1c1879e01ceee79a81693c67a1dba184b5fc1b1):


```
=== RUN   TestMultiRegionDataDriven/global_tables
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2704/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:1:
        new-cluster [1 args]
        <no input to command>
        ----
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2704/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:4:
        exec-sql [1 args]
        CREATE DATABASE db PRIMARY REGION "us-east-1" REGIONS  "us-west-1", "us-central-1";
        ----
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2704/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:8:
        exec-sql [1 args]
        CREATE TABLE db.global(k INT PRIMARY KEY, v INT) LOCALITY GLOBAL;
        ----
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2704/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:12:
        exec-sql [1 args]
        INSERT INTO db.global VALUES (1, 1), (2,2)
        ----
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #75543 pkg/ccl/multiregionccl/multiregionccl_test: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-master]
- #74783 ccl/multiregionccl: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-master]
- #73829 ccl/multiregionccl: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-release-21.2]
</p>
</details>
/cc @cockroachdb/multiregion
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMultiRegionDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-13833

ccl/multiregionccl.TestMultiRegionDataDriven [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4728289&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4728289&tab=artifacts#/) on release-22.1 @ [ff4e23e53388d3639f4b1b23b66e87d436556a8f](https://github.com/cockroachdb/cockroach/commits/ff4e23e53388d3639f4b1b23b66e87d436556a8f):


```
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3361/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:63:
        refresh-range-descriptor-cache [2 args]
        SELECT * FROM db.global WHERE k = 2
        ----
        LEAD_FOR_GLOBAL_READS
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3361/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:68:
        trace-sql [1 args]
        SELECT * FROM db.global WHERE k = 1
        ----
        served locally: true
        served via follower read: true
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3361/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:74:
        refresh-range-descriptor-cache [2 args]
        SELECT * FROM db.global WHERE k = 2
        ----
        LEAD_FOR_GLOBAL_READS
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3361/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:81:
        trace-sql [1 args]
        SELECT * FROM db.global WHERE k = 1
        ----
        served locally: false
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3361/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:86:
        exec-sql [1 args]
        ALTER DATABASE db ADD REGION "eu-central-1"
        ----
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3361/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:90:
        wait-for-zone-config-changes [6 args]
        <no input to command>
        ----
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3361/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:93:
        refresh-range-descriptor-cache [2 args]
        SELECT * FROM db.global WHERE k = 2
        ----
        LEAD_FOR_GLOBAL_READS
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3361/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:98: SELECT * FROM db.global WHERE k = 1
        expected:
        served locally: true
        served via follower read: true
        
        found:
        served locally: true
        served via follower read: false
    --- FAIL: TestMultiRegionDataDriven/global_tables (24.19s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #75543 pkg/ccl/multiregionccl/multiregionccl_test: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-master]
- #74783 ccl/multiregionccl: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-master]
- #73829 ccl/multiregionccl: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-release-21.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMultiRegionDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/multiregionccl.TestMultiRegionDataDriven [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4898004&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4898004&tab=artifacts#/) on release-22.1 @ [8c867d72534db2c64beb1a36bde9f8345bfd5e9f](https://github.com/cockroachdb/cockroach/commits/8c867d72534db2c64beb1a36bde9f8345bfd5e9f):


```
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3364/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:63:
        refresh-range-descriptor-cache [2 args]
        SELECT * FROM db.global WHERE k = 2
        ----
        LEAD_FOR_GLOBAL_READS
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3364/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:68:
        trace-sql [1 args]
        SELECT * FROM db.global WHERE k = 1
        ----
        served locally: true
        served via follower read: true
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3364/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:74:
        refresh-range-descriptor-cache [2 args]
        SELECT * FROM db.global WHERE k = 2
        ----
        LEAD_FOR_GLOBAL_READS
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3364/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:81:
        trace-sql [1 args]
        SELECT * FROM db.global WHERE k = 1
        ----
        served locally: false
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3364/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:86:
        exec-sql [1 args]
        ALTER DATABASE db ADD REGION "eu-central-1"
        ----
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3364/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:90:
        wait-for-zone-config-changes [6 args]
        <no input to command>
        ----
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3364/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:93:
        refresh-range-descriptor-cache [2 args]
        SELECT * FROM db.global WHERE k = 2
        ----
        LEAD_FOR_GLOBAL_READS
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3364/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:98: SELECT * FROM db.global WHERE k = 1
        expected:
        served locally: true
        served via follower read: true
        
        found:
        served locally: true
        served via follower read: false
    --- FAIL: TestMultiRegionDataDriven/global_tables (27.23s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #75543 pkg/ccl/multiregionccl/multiregionccl_test: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-master]
- #74783 ccl/multiregionccl: TestMultiRegionDataDriven failed [C-test-failure O-robot T-kv T-sql-experience branch-master]
- #73829 ccl/multiregionccl: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-release-21.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMultiRegionDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/multiregionccl.TestMultiRegionDataDriven [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5348471&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5348471&tab=artifacts#/) on release-22.1 @ [73ddb31f9b156f429f93a62cb9ce22da1fb2fe83](https://github.com/cockroachdb/cockroach/commits/73ddb31f9b156f429f93a62cb9ce22da1fb2fe83):


```
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3767/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:63:
        refresh-range-descriptor-cache [2 args]
        SELECT * FROM db.global WHERE k = 2
        ----
        LEAD_FOR_GLOBAL_READS
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3767/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:68:
        trace-sql [1 args]
        SELECT * FROM db.global WHERE k = 1
        ----
        served locally: true
        served via follower read: true
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3767/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:74:
        refresh-range-descriptor-cache [2 args]
        SELECT * FROM db.global WHERE k = 2
        ----
        LEAD_FOR_GLOBAL_READS
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3767/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:81:
        trace-sql [1 args]
        SELECT * FROM db.global WHERE k = 1
        ----
        served locally: false
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3767/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:86:
        exec-sql [1 args]
        ALTER DATABASE db ADD REGION "eu-central-1"
        ----
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3767/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:90:
        wait-for-zone-config-changes [6 args]
        <no input to command>
        ----
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3767/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:93:
        refresh-range-descriptor-cache [2 args]
        SELECT * FROM db.global WHERE k = 2
        ----
        LEAD_FOR_GLOBAL_READS
    datadriven_test.go:122: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3767/execroot/cockroach/bazel-out/k8-dbg/bin/pkg/ccl/multiregionccl/multiregionccl_test_/multiregionccl_test.runfiles/cockroach/pkg/ccl/multiregionccl/testdata/global_tables:98: SELECT * FROM db.global WHERE k = 1
        expected:
        served locally: true
        served via follower read: true
        
        found:
        served locally: true
        served via follower read: false
    --- FAIL: TestMultiRegionDataDriven/global_tables (26.23s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #80902 ccl/multiregionccl: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-release-22.1.0]
- #80837 ccl/multiregionccl: TestMultiRegionDataDriven failed [C-test-failure O-robot branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMultiRegionDataDriven.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/multiregionccl/datadriven_test.go
+++ b/pkg/ccl/multiregionccl/datadriven_test.go
@@ -210,6 +210,7 @@ SET CLUSTER SETTING kv.closed_timestamp.propagation_slack = '0.5s'
 
 			case "trace-sql":
 				mustHaveArgOrFatal(t, d, serverIdx)
+				var rec tracingpb.Recording
 				queryFunc := func() (localRead bool, followerRead bool, err error) {
 					var idx int
 					d.ScanArgs(t, serverIdx, &idx)
@@ -232,7 +233,7 @@ SET CLUSTER SETTING kv.closed_timestamp.propagation_slack = '0.5s'
 					if err != nil {
 						return false, false, err
 					}
-					rec := <-recCh
+					rec = <-recCh
 					localRead, followerRead, err = checkReadServedLocallyInSimpleRecording(rec)
 					if err != nil {
 						return false, false, err
@@ -250,6 +251,9 @@ SET CLUSTER SETTING kv.closed_timestamp.propagation_slack = '0.5s'
 				if localRead {
 					output.WriteString(
 						fmt.Sprintf("served via follower read: %s\n", strconv.FormatBool(followerRead)))
+				}
+				if d.Expected != output.String() {
+					return errors.AssertionFailedf("not a match, trace:\n%s\n", rec).Error()
 				}
 				return output.String()
 
@@ -333,8 +337,8 @@ SET CLUSTER SETTING kv.closed_timestamp.propagation_slack = '0.5s'
 						}
 					}
 
-					// If the user specified a leaseholder, transfer range lease to the
-					// leaseholder.
+					// If the user specified a leaseholder, transfer range's lease to the
+					// that node.
 					if expectedPlacement.hasLeaseholderInfo() {
 						expectedLeaseIdx := expectedPlacement.getLeaseholder()
 						actualLeaseIdx := actualPlacement.getLeaseholder()
@@ -365,7 +369,17 @@ SET CLUSTER SETTING kv.closed_timestamp.propagation_slack = '0.5s'
 							)
 						}
 					}
-
+					// Now that this range has gone through a bunch of changes, we lookup
+					// the range and its leaseholder again to ensure we're comparing the
+					// most up-to-date range state with the supplied expectation.
+					desc, err = ds.tc.LookupRange(lookupKey)
+					if err != nil {
+						return err
+					}
+					actualPlacement, err = parseReplicasFromRange(t, ds.tc, desc)
+					if err != nil {
+						return err
+					}
 					err = actualPlacement.satisfiesExpectedPlacement(expectedPlacement)
 					if err != nil {
 						return err
@@ -574,6 +588,7 @@ func parseReplicasFromInput(
 }
 
 // parseReplicasFromInput constructs a replicaPlacement from a range descriptor.
+// It also ensures all replicas have the same view of who the leaseholder is.
 func parseReplicasFromRange(
 	t *testing.T, tc serverutils.TestClusterInterface, desc roachpb.RangeDescriptor,
 ) (*replicaPlacement, error) {
@@ -585,6 +600,28 @@ func parseReplicasFromRange(
 	if err != nil {
 		return nil, errors.Wrap(err, "could not get leaseholder")
 	}
+	// This test performs lease transfers at various points and expects the
+	// range's leaseholder to conform to what was specified in the test. However,
+	// whenever the lease is transferred using methods on TestCluster, only the
+	// outgoing leaseholder is guaranteed to have applied the lease. Given the
+	// tracing assumptions these tests make, it's worth ensuring all replicas have
+	// applied the lease, as it makes reasoning about these tests much easier.
+	// To that effect, we loop through all replicas on the supplied descriptor and
+	// ensure that is indeed the case.
+	for _, repl := range desc.Replicas().VoterAndNonVoterDescriptors() {
+		t := tc.Target(nodeIdToIdx(t, tc, repl.NodeID))
+		lh, err := tc.FindRangeLeaseHolder(desc, &t)
+		if err != nil {
+			return nil, err
+		}
+		if lh.NodeID != leaseHolder.NodeID && lh.StoreID != leaseHolder.StoreID {
+			return nil, errors.Newf(
+				"all replicas do not have the same view of the lease; found %s and %s",
+				lh, leaseHolder,
+			)
+		}
+	}
+
 	leaseHolderIdx := nodeIdToIdx(t, tc, leaseHolder.NodeID)
 	replicaMap[leaseHolderIdx] = replicaTypeLeaseholder
 	ret.leaseholder = leaseHolderIdx
@@ -656,6 +693,12 @@ func (r *replicaPlacement) satisfiesExpectedPlacement(expected *replicaPlacement
 		}
 	}
 
+	if expected.hasLeaseholderInfo() && expected.getLeaseholder() != r.getLeaseholder() {
+		return errors.Newf(
+			"expected %s to be the leaseholder, but %s was instead",
+			expected.getLeaseholder(), r.getLeaseholder(),
+		)
+	}
 	return nil
 }
 

